### PR TITLE
Issue #254 refactor handle-resource-offer! and add unit tests

### PR DESCRIPTION
**Pull request by shamsimam**
Fixes https://github.com/twosigma/Cook/issues/254

### CODE REVIEW

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 1]
```diff
@@ -763,5 +766,385 @@
...
+  (testing "exceed quota limit - gpus"
+    (is (not (sched/below-quota? {:count 5, :cpus 15, :mem 9999, :gpus 4}
+                                 {:count 4, :cpus 10, :mem 1234, :gpus 5}))))
+  (testing "at quota limit with extra keys"
+    (is (sched/below-quota? {:count 5, :cpus 15, :mem 9999, :gpus 4}
+                            {:count 5, :cpus 15, :mem 9999}))))
+
+(deftest test-job->usage
+  (testing "cpus mem and gpus usage"
+    (is (= {:cpus 2, :mem 2048, :count 1} ;; TODO verify whether gpus are intentionally omitted?
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Does `job->usage` intentionally not include gpus in its report?

File: scheduler/src/cook/mesos/dru.clj [hunk 2]
```diff
@@ -53,7 +53,7 @@
 
 (defn compute-task-scored-task-pairs
   "Takes a sorted seq of task entities and dru-divisors, returns a list of [task scored-task], preserving the same order of input tasks"
-  [task-ents {mem-divisor :mem cpus-divisor :cpus}]
+  [{mem-divisor :mem cpus-divisor :cpus} task-ents]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/dru.clj]: reorder arguments to simplify use of arrow (`->>`) macro.

File: scheduler/src/cook/mesos/dru.clj [hunk 3]
```diff
@@ -69,20 +69,19 @@
...
-        task-cum-gpus (->> task-resources
-                           (accumulate-resources)
-                           (map (fn [{:keys [gpus]}]
-                                  (/ gpus gpu-divisor))))
-        scored-tasks (map (fn [task cum-gpus]
-                            [task cum-gpus])
-                          task-ents
-                          task-cum-gpus)]
-    scored-tasks))
+  [gpu-divisor task-ents]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/dru.clj]: reorder arguments to simplify use of arrow (`->>`) macro.

File: scheduler/src/cook/mesos/dru.clj [hunk 4]
```diff
@@ -108,26 +107,25 @@
...
+              (compute-sorted-task-cumulative-gpu-score-pairs (-> user user->dru-divisors :gpus) task-ents)))
        (sorted-merge second)))
 
 (timers/deftimer [cook-mesos dru sorted-task-scored-task-pairs-duration])
 
 (defn sorted-task-scored-task-pairs
   "Returns a lazy sequence of [task,scored-task] pairs sorted by dru in ascending order.
    If jobs have the same dru, any ordering is allowed"
-  [user->sorted-running-task-ents user->dru-divisors]
+  [user->dru-divisors user->sorted-running-task-ents]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/dru.clj]: reorder arguments to simplify use of arrow (`->>`) macro.

File: scheduler/src/cook/mesos/dru.clj [hunk 5]
```diff
@@ -108,26 +107,25 @@
   ([colls]
    (sorted-merge identity compare colls)))
 
-(defn gpu-task-scored-task-pairs
+(defn sorted-task-cumulative-gpu-score-pairs
   "Takes a sorted seq of task entities and the gpu divisor, returns a list of [task cumulative-gpus], preserving the same order of input tasks"
-  [user->sorted-running-task-ents user->dru-divisors]
+  [user->dru-divisors user->sorted-running-task-ents]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/dru.clj]: reorder arguments to simplify use of arrow (`->>`) macro.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 6]
```diff
@@ -14,47 +14,46 @@
 ;; limitations under the License.
 ;;
 (ns cook.mesos.scheduler
-  (:require [mesomatic.scheduler :as mesos]
-            [mesomatic.types :as mtypes]
+  (:require [chime :refer [chime-at chime-ch]]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: lexicographically ordered the import statements.

**Comment by shamsimam** [comment]
Passing build: https://travis-ci.org/twosigma/Cook/builds/211901342

File: scheduler/test/metatransaction/utils_test.clj [hunk 7]
```diff
@@ -81,175 +81,186 @@
 
 (deftest pass-test
   (testing "simple-pass-test"
-    (let [conn (get-test-conn)
-          update-out (update!! conn opts (fn [_] uno-txn))]
-      (is (:transaction update-out)))))
+    (with-out-str
```
Comment (shamsimam) [inline scheduler/test/metatransaction/utils_test.clj]: Suppress console output while running unit tests

**Comment by dposada** [comment]
Can we squash and remove the `--` from commit message?

File: scheduler/test/cook/test/mesos/dru.clj [hunk 8]
```diff
@@ -89,8 +89,64 @@
             ordered-drus [1.0 1.0 1.0 1.5 4.0 5.5]]
         (is (= ordered-drus
                (map (comp :dru second)
-                    (dru/sorted-task-scored-task-pairs (map-vals (partial sort-by identity (util/same-user-task-comparator))
-                                                          (group-by util/task-ent->user task-ents))
-                                                {"ljin" share "wzhao" share "sunil" share}))))))))
+                    (dru/sorted-task-scored-task-pairs
+                      {"ljin" share "wzhao" share "sunil" share}
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/dru.clj]: How about "user1", "user2", "user3" instead of "ljin", "wzhao", "sunil"?

File: scheduler/test/cook/test/mesos/dru.clj [hunk 9]
```diff
@@ -89,8 +89,64 @@
...
+    (let [datomic-uri "datomic:mem://test-score-tasks"
+          conn (restore-fresh-database! datomic-uri)
+          job1 (create-dummy-job conn :user "ljin" :gpus 10.0)
+          job2 (create-dummy-job conn :user "ljin" :gpus 5.0)
+          job3 (create-dummy-job conn :user "ljin" :gpus 25.0)
+          job4 (create-dummy-job conn :user "ljin" :gpus 15.0)
+          task1 (create-dummy-instance conn job1 :instance-status :instance.status/running)
+          task2 (create-dummy-instance conn job2 :instance-status :instance.status/running)
+          task3 (create-dummy-instance conn job3 :instance-status :instance.status/running)
+          task4 (create-dummy-instance conn job4 :instance-status :instance.status/running)
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/dru.clj]: Should we extract a function for 

```clj
(create-dummy-instance conn % :instance-status :instance.status/running)
``` 
since it's used several times?

File: scheduler/test/cook/test/mesos/dru.clj [hunk 10]
```diff
@@ -89,8 +89,64 @@
...
+          task-ent1 (d/entity (d/db conn) task1)
+          task-ent2 (d/entity (d/db conn) task2)
+          task-ent3 (d/entity (d/db conn) task3)
+          task-ent4 (d/entity (d/db conn) task4)
+          tasks [task-ent1 task-ent2 task-ent3 task-ent4]]
+      (let [expected-result [[task-ent1 1.0] [task-ent2 1.5] [task-ent3 4.0] [task-ent4 5.5]]]
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/dru.clj]: The `let` seems unnecessary here

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 11]
```diff
@@ -763,5 +766,385 @@
       @(sched/receive-offers offers-chan mock-driver [offer-3])
       (is (= @declined-offer-ids-atom [(:id offer-2) (:id offer-3)])))))
 
+(deftest test-below-quota?
+  (testing "not using quota"
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/scheduler.clj]: I would favor a newline before each `testing` clause for readability

**Comment by shamsimam** [comment]
@dposada I will squash at the end of the review before merging. Removed the `--` from the commit messages.

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 12]
```diff
@@ -763,5 +766,385 @@
...
+             (sched/filter-based-on-quota {test-user {:count 10, :cpus 20, :mem 32768}} user->usage queue))))
+    (testing "room for later jobs not included"
+      (is (= [(make-job 1 2 2048) (make-job 2 1 1024)]
+             (sched/filter-based-on-quota {test-user {:count 4, :cpus 20, :mem 6144}} user->usage queue))))))
+
+(deftest test-category->pending-jobs->category->considerable-jobs
+  (let [uri "datomic:mem://test-category-pending-jobs-category-considerable-jobs"
+        conn (restore-fresh-database! uri)
+        test-db (d/db conn)
+        test-user (System/getProperty "user.name")
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Do we have to use the `user.name` property here?

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 1 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: @wyegelwel @DiegoAlbertoTorres do you guys know the answer?

File: scheduler/test/cook/test/mesos/dru.clj (hunk 8 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/dru.clj]: These are values from a pre-existing test. I didn't want to change the names used in this test as I just reordered the arguments. We can do that as a separate pass for all tests.

File: scheduler/test/cook/test/mesos/dru.clj (hunk 9 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/dru.clj]: I'm okay leaving the function at `create-dummy-instance` level as it is now.

File: scheduler/test/cook/test/mesos/dru.clj (hunk 10 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/dru.clj]: Will remove.

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 11 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: I felt the testing blocks were sparse, so didn't introduce new lines.

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 12 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Yes, `create-dummy-job` uses `user.name` while creating the jobs.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 6 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Let's get a pass on this file as it has already been done. I will create a separate PR for all the other files together.

File: scheduler/src/cook/mesos/dru.clj [hunk 13]
```diff
@@ -69,20 +69,19 @@
...
-                           (accumulate-resources)
-                           (map (fn [{:keys [gpus]}]
-                                  (/ gpus gpu-divisor))))
-        scored-tasks (map (fn [task cum-gpus]
-                            [task cum-gpus])
-                          task-ents
-                          task-cum-gpus)]
-    scored-tasks))
+  [gpu-divisor task-ents]
+  (if (seq task-ents)
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/dru.clj]: Why add this check here? If task-ents is empty all of these maps will should be effective no-ops

File: scheduler/src/cook/mesos/scheduler.clj [hunk 14]
```diff
@@ -486,8 +487,9 @@
   ;; The select-keys on quota was added because if there is a
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/scheduler.clj]: You can remove this comment given the code improvement

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 15]
```diff
@@ -822,5 +828,200 @@
...
+(deftest test-handle-resource-offers
+  (let [uri "datomic:mem://test-handle-resource-offers"
+        launched-offer-ids-atom (atom [])
+        launched-task-ids-atom (atom [])
+        driver (reify msched/SchedulerDriver
+                 (launch-tasks! [_ offer-id tasks]
+                   (swap! launched-offer-ids-atom conj (-> offer-id first :value))
+                   (swap! launched-task-ids-atom concat (map (fn [task]
+                                                               {:id (get-in task [:task-id :value])
+                                                                :cpus (->> task
```
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Why no `:mem` here?

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 16]
```diff
@@ -822,5 +828,200 @@
       (is (= [(make-job 1 2 2048) (make-job 2 1 1024)]
              (sched/filter-based-on-quota {test-user {:count 4, :cpus 20, :mem 6144}} user->usage queue))))))
 
+(deftest test-handle-resource-offers
```
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/scheduler.clj]: https://media.giphy.com/media/MziKDo6gO7x8A/giphy.gif

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 15 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: I'm using prime numbers for the cpus and validating their sum is what is matched. I'll do soemthing similar with `:mem`.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 17]
```diff
@@ -522,158 +529,179 @@
...
 
+(defn- category->pending-jobs->category->considerable-jobs
+  "Limit the pending jobs to considerable jobs based on usage and quota.
+   Further limit the considerable jobs to a maximum of num-considerable jobs."
+  [db pending-jobs user->quota user->usage num-considerable]
+  (log/debug "There are" (apply + (map count pending-jobs)) "pending jobs")
+  (log/debug "pending-jobs:" pending-jobs)
+  (let [category->considerable-jobs
+        (->> pending-jobs
+             (map (fn [[category jobs]]
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/scheduler.clj]: The interesting bit here is the filters on jobs. I propose putting that into a function called something like `filter-to-considerable` (or, you know, something way better) 

```clojure
(->> jobs
       ....)
```

Then this let block can become:

```clojure
(let [category->considerable-jobs (->> pending-jobs
                                  (map-vals filter-to-considerable))]
    ...)
```

File: scheduler/src/cook/mesos/scheduler.clj [hunk 18]
```diff
@@ -522,158 +529,179 @@
...
+     :matched-normal-job-uuids matched-normal-job-uuids}))
+
+(defn- remove-matched-jobs-from-pending-jobs
+  "Removes matched normal and gpu jobs from category->pending-jobs."
+  [category->pending-jobs matched-normal-job-uuids matched-gpu-job-uuids]
+  (let [remove-matched-jobs (fn remove-matched-jobs [existing-jobs matched-job-uuids]
+                              (remove #(contains? matched-job-uuids (:job/uuid %)) existing-jobs))
+        update-scheduler-contents (fn update-scheduler-contents [category->pending-jobs]
+                                    (-> category->pending-jobs
+                                        (update-in [:gpu] remove-matched-jobs matched-gpu-job-uuids)
```
Comment (wyegelwel) [inline scheduler/src/cook/mesos/scheduler.clj]: you can use `update` and remove the `[` and `]` from `[:gpu]`

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 15 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/scheduler.clj]: That is fine then. Just put a comment why mem is missing

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 1 above)
Comment (wyegelwel) [inline scheduler/test/cook/test/mesos/scheduler.clj]: its a bug, gpu should be included

Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Okay I will create a separate issue and PR for it.

Created: https://github.com/twosigma/Cook/issues/262

File: scheduler/src/cook/mesos/dru.clj (hunk 13 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/dru.clj]: I kept the behavior and structure same as `compute-task-scored-task-pairs` which has the same if branching.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 18 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done: https://github.com/twosigma/Cook/pull/259/commits/72ff7ca885388a9f7c252e741055e442c4108952

File: scheduler/src/cook/mesos/scheduler.clj (hunk 17 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done: https://github.com/twosigma/Cook/pull/259/commits/72ff7ca885388a9f7c252e741055e442c4108952

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 15 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Done: https://github.com/twosigma/Cook/pull/259/commits/72ff7ca885388a9f7c252e741055e442c4108952

File: scheduler/src/cook/mesos/scheduler.clj (hunk 14 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done: https://github.com/twosigma/Cook/pull/259/commits/72ff7ca885388a9f7c252e741055e442c4108952

**Comment by shamsimam** [comment]
Rebased the commits to handle merge conflicts.

**Comment by shamsimam** [comment]
Ready for next phase of reviews :)

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 19]
```diff
@@ -763,5 +766,386 @@
...
+        (is (= {:normal [job-1], :gpu [job-5]}
+               (sched/category->pending-jobs->category->considerable-jobs
+                 (d/db conn) category->pending-jobs user->quota user->usage num-considerable)))))
+    (testing "all jobs exceed quota"
+      (let [user->usage {test-user {:count 1, :cpus 2, :mem 1024, :gpus 0}}
+            user->quota {test-user {:count 5, :cpus 3, :mem 4096, :gpus 10}}
+            num-considerable 5]
+        (is (= {:normal [], :gpu []}
+               (sched/category->pending-jobs->category->considerable-jobs
+                 (d/db conn) category->pending-jobs user->quota user->usage num-considerable)))))))
```
Comment (DiegoAlbertoTorres) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Can we add a test where num-considerable actually truncates the amount of considerable jobs?

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 20]
```diff
@@ -763,5 +766,386 @@
...
+                                        (async/>!! offers-chan :end-marker)
+                                        result))]
+
+    (testing "enough offers for all normal jobs"
+      (let [num-considerable 10
+            offers [offer-1 offer-2 offer-3]]
+        (is (run-handle-resource-offers! num-considerable offers))
+        (is (= :end-marker (async/<!! offers-chan)))
+        (is (= 3 (count @launched-offer-ids-atom)))
+        (is (= (+ 3 13 7 11) (reduce + (map :cpus @launched-task-ids-atom))))
```
Comment (DiegoAlbertoTorres) [inline scheduler/test/cook/test/mesos/scheduler.clj]: You can set a job-name with create-dummy job and then get the name from a task (`(get-in task [:job :job/name])`). Then you can use the job name to identify jobs in your assertions, instead of using prime amounts of CPUs as identifiers.

Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Will do, thanks.

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 19 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Will do, already have a similar one in `test-handle-resource-offers`

File: scheduler/src/cook/mesos/scheduler.clj [hunk 21]
```diff
@@ -521,158 +520,178 @@
...
         (catch Throwable t
           (meters/mark! handle-resource-offer!-errors)
           (log/error t "Error in match:" (ex-data t))
           (when-let [offers @offer-stash]
             (async/go
               (async/>! offers-chan offers)))
-          true  ; if an error happened, it doesn't mean we need to penalize Fenzo
-          )))))
+          ; if an error happened, it doesn't mean we need to penalize Fenzo
+          true)))))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: He he.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 22]
```diff
@@ -521,158 +520,178 @@
...
+    (log/debug "matched normal jobs:" (count matched-normal-job-uuids))
+    (log/debug "matched gpu jobs:" (count matched-gpu-job-uuids))
+    (when (not (empty? matches))
+      (let [matched-normal-jobs-resource-requirements (util/sum-resources-of-jobs matched-normal-jobs)]
+        (meters/mark! matched-tasks-cpus (:cpus matched-normal-jobs-resource-requirements))
+        (meters/mark! matched-tasks-mem (:mem matched-normal-jobs-resource-requirements))))
+    {:matched-gpu-job-uuids matched-gpu-job-uuids
+     :matched-normal-job-uuids matched-normal-job-uuids}))
+
+(defn remove-matched-jobs-from-pending-jobs
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Will do.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 23]
```diff
@@ -521,158 +520,178 @@
...
+                                        (filter (fn [job]
+                                                  (util/job-allowed-to-start? db job)))
+                                        (take num-considerable)))
+        category->considerable-jobs (->> category->pending-jobs
+                                         (map-vals filter-considerable-jobs))]
+    (log/debug "We'll consider scheduling" (map (fn [[k v]] [k (count v)]) category->considerable-jobs)
+               "of those pending jobs (limited to " num-considerable " due to backdown)")
+    category->considerable-jobs))
+
+(defn extract-matched-job-uuids
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Will do. The beauty is now that we have unit tests, easy to validate if something still works the same :)

File: scheduler/src/cook/mesos/scheduler.clj [hunk 24]
```diff
@@ -521,158 +520,178 @@
...
 
+(defn category->pending-jobs->category->considerable-jobs
+  "Limit the pending jobs to considerable jobs based on usage and quota.
+   Further limit the considerable jobs to a maximum of num-considerable jobs."
+  [db category->pending-jobs user->quota user->usage num-considerable]
+  (log/debug "There are" (apply + (map count category->pending-jobs)) "pending jobs")
+  (log/debug "pending-jobs:" category->pending-jobs)
+  (let [filter-considerable-jobs (fn filter-considerable-jobs [jobs]
+                                   (->> jobs
+                                        (map #(d/entity db (:db/id %)))
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: `jobs` contains the `:db/id`, which is why we extract the id using the call to `(:db/id %)`. I like the naming strategy of using `*job-ids` though.

Comment (DiegoAlbertoTorres) [inline scheduler/src/cook/mesos/scheduler.clj]: Ah, right, I see. Is it just pulling from the database to get updated information then?

Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Yes

**Comment by wyegelwel** [comment]
Can you add timing calls around each of the sub pieces of handle resource offers so we can start to better understand where the time is going?

**Comment by shamsimam** [comment]
@wyegelwel  will add the timing calls.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 23 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Ended up using this:
```
(let [category->jobs (group-by util/categorize-job
                                 (->> matches
                                      (map #(-> % :tasks))
                                      flatten
                                      (map #(-> % .getRequest :job))))
        category->job-uuids (map-vals #(set (map :job/uuid %)) category->jobs)]
    (log/debug "matched jobs:" (map-vals count category->job-uuids))
    (when (not (empty? matches))
      (let [matched-normal-jobs-resource-requirements (-> category->jobs :normal util/sum-resources-of-jobs)]
        (println matched-normal-jobs-resource-requirements) ;; todo remove
        (meters/mark! matched-tasks-cpus (:cpus matched-normal-jobs-resource-requirements))
        (meters/mark! matched-tasks-mem (:mem matched-normal-jobs-resource-requirements))))
    category->job-uuids)
```

File: scheduler/src/cook/mesos/scheduler.clj (hunk 22 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done.

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 19 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Done.

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 20 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Done.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 23 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: https://github.com/twosigma/Cook/pull/259/commits/b2c85b4d4529ddd9aa6c6cd122e38b9108936a20

**Comment by shamsimam** [comment]
Ready for another phase of reviews :)

**Comment by wyegelwel** [comment]
Please rebase when you get a chance

**Comment by shamsimam** [comment]
rebased.

File: scheduler/src/cook/mesos/scheduler.clj [hunk 25]
```diff
@@ -524,158 +525,174 @@
...
+(defn matches->category->job-uuids
+  "Returns the matched job uuid sets by category."
+  [matches]
+  (let [category->jobs (group-by util/categorize-job
+                                 (->> matches
+                                      (map #(-> % :tasks vec))
+                                      flatten
+                                      (map #(-> % .getRequest :job))))
+        category->job-uuids (map-vals #(set (map :job/uuid %)) category->jobs)]
+    (log/debug "matched jobs:" (map-vals count category->job-uuids))
```
Comment (dposada) [inline scheduler/src/cook/mesos/scheduler.clj]: Why are we using `map-vals` without `pc/`? Elsewhere we use e.g. `pc/map-from-keys`

File: scheduler/src/cook/mesos/scheduler.clj [hunk 26]
```diff
@@ -524,158 +525,174 @@
...
+                      :instance/hostname (.getHostname task)
+                      :instance/start-time (now)
+                      ;; NB command executor uses the task-id
+                      ;; as the executor-id
+                      :instance/executor-id task-id
+                      :instance/slave-id slave-id
+                      :instance/ports (.getAssignedPorts task)
+                      :instance/progress 0
+                      :instance/status :instance.status/unknown
+                      :instance/preempted? false}])]
```
Comment (dposada) [inline scheduler/src/cook/mesos/scheduler.clj]: Since this is a refactoring, it seems to me like this `for` is a good candidate for extracting into a function, perhaps:

```clj
(defn- task-txns [matches]
  (for ...))
```

File: scheduler/src/cook/mesos/scheduler.clj [hunk 27]
```diff
@@ -524,158 +525,174 @@
...
+    (meters/mark! scheduler-offer-matched
+                  (->> matches
+                       (mapcat (comp :id :offer :leases))
+                       (distinct)
+                       (count)))
+    (histograms/update! number-offers-matched
+                        (->> matches
+                             (mapcat (comp :id :offer :leases))
+                             (distinct)
+                             (count)))
```
Comment (dposada) [inline scheduler/src/cook/mesos/scheduler.clj]: This code is duplicated for the two metrics above:

```clj
(->> matches
  (mapcat (comp :id :offer :leases))
  (distinct)
  (count))
```

File: scheduler/src/cook/mesos/scheduler.clj [hunk 28]
```diff
@@ -1027,16 +1044,14 @@
...
-                     (filter (fn [[task _]]
-                               (contains? pending-task-ents task)))
-                     (map (fn [[task _]]
-                            (:job/_instance task)))))]
+               (->> (concat running-task-ents pending-task-ents)
+                    (group-by util/task-ent->user)
+                    (map-vals (fn [task-ents] (into (sorted-set-by (util/same-user-task-comparator)) task-ents)))
+                    (dru/sorted-task-cumulative-gpu-score-pairs user->dru-divisors)
+                    (filter (fn [[task _]]
+                              (contains? pending-task-ents task)))
```
Comment (dposada) [inline scheduler/src/cook/mesos/scheduler.clj]: Nitpick: can leave this on previous line

File: scheduler/src/cook/mesos/scheduler.clj [hunk 29]
```diff
@@ -1027,16 +1044,14 @@
...
+                    (map (fn [[task _]]
+                           (:job/_instance task)))))]
```
Comment (dposada) [inline scheduler/src/cook/mesos/scheduler.clj]: Nitpick: can leave this on previous line

File: scheduler/test/cook/test/mesos/dru.clj (hunk 8 above)
Comment (dposada) [inline scheduler/test/cook/test/mesos/dru.clj]: OK

File: scheduler/test/cook/test/mesos/dru.clj [hunk 30]
```diff
@@ -89,8 +89,63 @@
...
+                      {"ljin" share "wzhao" share "sunil" share}
+                      (map-vals (partial sort-by identity (util/same-user-task-comparator))
+                                (group-by util/task-ent->user task-ents))))))))))
+
+(deftest test-compute-sorted-task-cumulative-gpu-score-pairs
+  (testing "return empty set on input empty set"
+    (is (= []
+           (dru/compute-sorted-task-cumulative-gpu-score-pairs 25.0 '()))))
+
+  (testing "test1"
```
Comment (dposada) [inline scheduler/test/cook/test/mesos/dru.clj]: What is "test1"?

File: scheduler/test/cook/test/mesos/scheduler.clj (hunk 1 above)
Comment (dposada) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Can we remove the TODO?

File: scheduler/src/cook/mesos/scheduler.clj (hunk 25 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: I prefer prefixing with `pc/` too. However, `map-vals` was already referred explicitly and I did not want to introduce another round of refactoring in this file. Maybe, we can do that cleanup as part of another PR.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 26 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 27 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 28 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 29 above)
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Done.

File: scheduler/test/cook/test/mesos/dru.clj (hunk 30 above)
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/dru.clj]: Used proper explanation

**Comment by shamsimam** [comment]
rebased. addressed feedback from previous round.

File: scheduler/src/cook/mesos/scheduler.clj (hunk 25 above)
Comment (dposada) [inline scheduler/src/cook/mesos/scheduler.clj]: That sounds good

**Comment by dposada** [comment]
LGTM
